File: java/LongestValidParentheses.java
public class LongestValidParentheses {
    public int longestValidParentheses(String s) {
        if (s == null) {
            return 0;
        }
        int n = s.length();
        if (n < 2) {
            return 0;
        }
        int[] stack = new int[n];
        int top = 0;
        int[] dp = new int[n + 1];
        if (s.charAt(0) == '(') {
            stack[top++] = 1;
        }
        for (int i = 1; i < n; ++i) {
            if (s.charAt(i) == '(') {
                stack[top++] = i + 1;
            } else {
                if (top != 0) {
                    int len = i - (stack[top - 1] - 1) + 1;
                    dp[i + 1] = dp[stack[top - 1] - 1] + len;
                    --top;
                }
            }
        }
        int longest = 0;
        for (int i = 0; i <= n; ++i) {
            longest = Math.max(longest, dp[i]);
        }
        return longest;
    }

    public static void main(String[] args) {
        LongestValidParentheses solution = new LongestValidParentheses();
        String s = null;

        s = "()(()";
        System.out.println(solution.longestValidParentheses(s));

        s = ")()())";
        System.out.println(solution.longestValidParentheses(s));

        s = "(()";
        System.out.println(solution.longestValidParentheses(s));
    }
}

File: cc/longest_valid_parentheses.cc
#include "base_include.h"

class LongestValidParentheses {
public:
    int longestValidParentheses(string s) {
        int n = s.size();
        if (0 == n) {
            return 0;
        }
        int result = 0;
        stack<int> S;
        vector<int> dp(n + 1, 0);
        if ('(' == s[0]) {
            S.push(1);
        }
        for (int i = 1; i < n; ++i) {
            if ('(' == s[i]) {
                S.push(i + 1);
            } else {
                if (!S.empty()) {
                    int len = i - S.top() + 2;
                    dp[i + 1] = dp[S.top() - 1] + len;
                    S.pop();
                }
            }
        }
        for (int i = 0; i <= n; ++i) {
            result = max(result, dp[i]);
        }
        return result;
    }
};

int main(int argc, char *argv[]) {
    LongestValidParentheses solver;
    string s;

    s = "()(())";
    cout << solver.longestValidParentheses(s) << endl;
    return 0;
}
